src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,220 @@
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\Doctrine;
+
+use Doctrine\DBAL\Connection as DriverConnection;
+use Doctrine\DBAL\Exception\TableNotFoundException;
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\DBAL\Schema\Synchronizer\SingleDatabaseSynchronizer;
+use Doctrine\DBAL\Types\Type;
+use Symfony\Component\Messenger\Exception\InvalidArgumentException;
+
+/**
+ * @author Vincent Touzet <vincent.touzet@gmail.com>
+ */
+class Connection
+{
+    /**
+     * Configuration of the connection.
+     *
+     * * table_name : name of the table
+     * * connection : name of the Doctrine's entity manager
+     * * redeliver_timeout : Timeout before redeliver messages still in handling state (i.e: delivered_at is not null and message is still in table). Default 3600
+     * * loop_sleep : Number of micro seconds to wait for a next message to handle
+     * * auto_setup : Whether the table should be created automatically during send / get. Default : true
+     */
+    private $configuration = [];
+    private $driverConnection;
+
+    public function __construct(array $configuration, DriverConnection $driverConnection)
+    {
+        $this->configuration = array_replace_recursive([
+            'table_name' => 'messenger_messages',
+            'queue' => 'default',
+            'redeliver_timeout' => 3600,
+            'loop_sleep' => 200000,
+            'auto_setup' => true,
+        ], $configuration);
+        $this->driverConnection = $driverConnection;
+    }
+
+    public function getConfiguration(): array
+    {
+        return $this->configuration;
+    }
+
+    public static function buildConfiguration($dsn, array $options = [])
+    {
+        if (false === $parsedUrl = parse_url($dsn)) {
+            throw new InvalidArgumentException(sprintf('The given Doctrine DSN "%s" is invalid.', $dsn));
+        }
+
+        $components = parse_url($dsn);
+        $query = [];
+        if (isset($components['query'])) {
+            parse_str($components['query'], $query);
+        }
+
+        return [
+            'connection' => $components['host'],
+            'table_name' => $options['table_name'] ?? ($query['table_name'] ?? 'messenger_messages'),
+            'queue' => $options['queue'] ?? ($query['queue'] ?? 'default'),
+            'redeliver_timeout' => $options['redeliver_timeout'] ?? ($query['redeliver_timeout'] ?? 3600),
+            'loop_sleep' => $options['loop_sleep'] ?? ($query['loop_sleep'] ?? 200000),
+            'auto_setup' => $options['auto_setup'] ?? ($query['auto_setup'] ?? true),
+        ];
+    }
+
+    public function send(string $body, array $headers, int $delay = 0): void
+    {
+        $availableAt = \DateTime::createFromFormat('U.u', microtime(true) + $delay / 1000);
```

Comment:
I'd move `$now` creation before this line (without the format) and simply add `$delay` to have the `$availableAt`. It would be simpler to read and would ensure the difference between `created_at` and `available_at` is exactly the expected delay.